### DISCUSSION TITLE
No unit/regression testing for the project?

### DISCUSSION THREAD

**Discussion by mgcrea** [TanStack/router#1304] [Q&A]
I have been using this router in my latest app and closely tracked the new releases, after hitting some kind of regression due to changes in the routeTree generation, I dug a little bit in the code and was quite surprised to see almost no unit testing in the whole project.

I find it a bit odd for a project that emphasise "safety" like this one, and looking at the high volume of releases (which I do like!) with a few reverts and no unit-testing whatsoever feels a bit unstable to me, especially for code that does not look like to be trivial with quite a few edge cases.

I'm not blaming anyone, it's open source and I thank you for that, but I'm honestly curious about the reasoning or difficulties that lead to this lack of automated testing.

Might try to give a hand if help is needed on this side.

**Comment by tannerlinsley** [comment]
Oh no! You found my weakness! 🫣 I feel your frustration. TS has been a great seatbelt thus far but we definitely have passed the point where we should have had plenty of tests set up. They could have already saved me! I would love help in this arena. Most of our other libraries are well tested so I’m fine if we rip their testing frameworks for Router. I’ll definitely need help though, since testing is not my strong suit.
